### PR TITLE
[mlir][SCF] Bufferize scf.index_switch

### PR DESCRIPTION
**Pull request by matthias-springer**
Add the `BufferizableOpInterface` implementation of `scf.index_switch`.

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 1]
```diff
@@ -295,6 +295,117 @@ struct IfOpInterface
...
+    auto switchOp = cast<scf::IndexSwitchOp>(op);
+    int64_t resultNum = cast<OpResult>(value).getResultNumber();
+    AliasingOpOperandList result;
+    for (int64_t i = 0, numCases = switchOp.getNumCases(); i < numCases; ++i) {
+      auto yieldOp =
+          cast<scf::YieldOp>(switchOp.getCaseBlock(i).getTerminator());
+      result.addAlias(AliasingOpOperand(&yieldOp->getOpOperand(resultNum),
+                                        BufferRelation::Equivalent,
+                                        /*isDefinite=*/false));
+    }
```
Comment (maerhart) [inline mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp:317]: Don't we have to consider the default region as well?

File: mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 2]
```diff
@@ -295,6 +295,117 @@ struct IfOpInterface
...
+      if (failed(bufferType))
+        return failure();
+      newTypes.push_back(*bufferType);
+    }
+
+    // Create new op.
+    rewriter.setInsertionPoint(switchOp);
+    auto newSwitchOp = rewriter.create<scf::IndexSwitchOp>(
+        switchOp.getLoc(), newTypes, switchOp.getArg(), switchOp.getCases(),
+        switchOp.getCases().size());
```
Comment (maerhart) [inline mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp:343]: Any attached attributes would get lost here. Shouldn't we add them to the new op again?

File: mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp [hunk 3]
```diff
@@ -295,6 +295,117 @@ struct IfOpInterface
...
+    replaceOpWithBufferizedValues(rewriter, op, newSwitchOp->getResults());
+
+    return success();
+  }
+
+  FailureOr<BaseMemRefType>
+  getBufferType(Operation *op, Value value, const BufferizationOptions &options,
+                SmallVector<Value> &invocationStack) const {
+    auto switchOp = cast<scf::IndexSwitchOp>(op);
+    assert(value.getDefiningOp() == op && "invalid valid");
```
Comment (maerhart) [inline mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp:363]: Should the message be "invalid operation"?

**Review by maerhart — APPROVED** [review]
Looks good.
Tests could be a bit more thorough, e.g.,
* Alias analysis test
* test where we need to switch to a dynamic layout

File: mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp (hunk 2 above)
Comment (matthias-springer) [reply mlir/lib/Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp:343]: We generally do not preserve custom attributes. Most transformations just drop them.

**Comment by matthias-springer** [comment]
> Looks good. Tests could be a bit more thorough, e.g.,
> 
> * Alias analysis test
> * test where we need to switch to a dynamic layout

Alias set tests are bit tedious to write, so I currently just add them when testing new functionality in the analysis but not on every op.

The first case switches to dynamic layout.

**Pull request merged**